modules/cudastereo/src/cuda/stereosgm_path_aggregation.cu

```diff
@@ -0,0 +1,96 @@
+/*
+Copyright 2016 Fixstars Corporation
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http ://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "path_aggregation.hpp"
+#include "vertical_path_aggregation.hpp"
+#include "horizontal_path_aggregation.hpp"
+#include "oblique_path_aggregation.hpp"
+
+namespace cv { namespace cuda { namespace device
+{
+    namespace path_aggregation
+    {
+template <size_t MAX_DISPARITY>
+PathAggregation<MAX_DISPARITY>::PathAggregation()
+	: m_cost_buffer()
+{
+	for(unsigned int i = 0; i < NUM_PATHS; ++i){
+		cudaStreamCreate(&m_streams[i]);
+		cudaEventCreate(&m_events[i]);
+	}
+}
+
+template <size_t MAX_DISPARITY>
+PathAggregation<MAX_DISPARITY>::~PathAggregation(){
+	for(unsigned int i = 0; i < NUM_PATHS; ++i){
+		cudaStreamSynchronize(m_streams[i]);
+		cudaStreamDestroy(m_streams[i]);
+		cudaEventDestroy(m_events[i]);
+	}
+}
+
+template <size_t MAX_DISPARITY>
+void PathAggregation<MAX_DISPARITY>::enqueue(
+	const feature_type *left,
+	const feature_type *right,
+	int width,
+	int height,
+	unsigned int p1,
+	unsigned int p2,
+	cudaStream_t stream)
+{
+	const size_t buffer_size = width * height * MAX_DISPARITY * NUM_PATHS;
+	if(m_cost_buffer.size() != buffer_size){
+		m_cost_buffer = DeviceBuffer<cost_type>(buffer_size);
+	}
+	const size_t buffer_step = width * height * MAX_DISPARITY;
+	cudaStreamSynchronize(stream);
+	path_aggregation::enqueue_aggregate_up2down_path<MAX_DISPARITY>(
+		m_cost_buffer.data() + 0 * buffer_step,
+		left, right, width, height, p1, p2, m_streams[0]);
+	path_aggregation::enqueue_aggregate_down2up_path<MAX_DISPARITY>(
```

Comment:
In Path Aggregation, destination data has four-dimensions that include NUM_PATHS, MAX_DISPARITY, height and width.   
We'd like to use OpenCV andor well-tested data structures as much as possible, on the other hand, we would not like to fail original code (performance).

What the best type do we choose?

1. thrust::device_vector
2. array of array of cv::cuda::GpuMat
3. raw pointer of arrary (bacause of internal use)

Reply:
4. GpuMat of GpuMat

In following internal implementation,  
it seems to be converting `std::vector<cv::cuda::GpuMat>` to GpuMat which has 1d pointers.  
https://github.com/opencv/opencv_contrib/blob/4.0.1/modules/cudafeatures2d/src/brute_force_matcher.cpp#L130

As applying this, convert data to GpuMat which has 2d pointers.

5. 1d GpuMat as 4d

LGTM.  I'm not sure but I feel strange since GpuMat provides pitched 2d array.


Then, I will implement no. 4 and 5, and more if I can, check performances respectively.